[OptimistLabyrinth] 🚀 1단계 - 엔티티 매핑

OptimistLabyrinth opened the pull request:
Q&A 서비스와 관련 있는
네 개의 엔티티 Answer, DeleteHistory, Question, User 클래스를 정의한 후
엔티티 정의가 정상적으로 이루어졌다는 것을
```@DataJpaTest``` 사용한 save( ) 메서드에 대한 테스트 케이스 작성을 통해서
개별 엔티티 객체가 데이터베이스에 저장되는 것을 확인했습니다.

Commit:
chore: 'assertj' 패키지 최신버전으로 업그레이드

Commit:
feat: 네 개의 JPA 엔티티 정의

Answer, DeleteHistory, Question, User 클래스 정의

Commit:
feat: 개별 엔티티에 대한 DataJpaTest 클래스 추가

개별 엔티티의 createdAt 필드를 세팅하기 위한 ```@PrePersist```,
	updatedAt 필드를 세팅하기 위한 ```@PreUpdate``` 메서드 정의

save() 메서드에 대한 테스트 코드 작성

mskangg commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,16 +1,36 @@
+@Entity
+@Table(name = "answer")
+@Getter
+@NoArgsConstructor
```
기본생성자를 public으로 열어두셨네요!
이러면 몇가지 문제점이 발생할 수 있는데요!
어떤것들이 있을까요? 🤔

OptimistLabyrinth replied:
정확하게 잘 아는건 아니지만 
JPA 내부적으로 NoArgsConstructor 를 호출한 뒤 멤버 변수를 하나씩 세팅하는 로직이 있다고 알고 있습니다.
그래서 `@NoArgsConstructor` 를 사용했어요.

그러나 기본생성자가 public 이면 Entity 클래스 객체를 생성해서 사용하는 경우가 많은 곳
예를 들면 서비스 레이어같은 곳에서
개발자가 실수로 NoArgsConstructor 를 호출해서 
유효하지 않은(invalid) 도메인 엔티티 객체를 만들어서 사용해서 
의도치 않은 오류가 발생할 수 있을 것 같아요.

OptimistLabyrinth replied:
수정 완료했습니다.

mskangg commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -32,44 +52,22 @@ public Answer(Long id, User writer, Question question, String contents) {
-        this.writerId = writerId;
+    @PrePersist
+    void onCreate() {
+        createdAt = LocalDateTime.now();
```
순수 jpa auditing 을 사용하셨군요 👍 
미션 중 전체적으로 공통되는 부분이니 abstract class를 만들고 spring data jpa auditing을 사용해 보는 것도 좋아보입니다.

OptimistLabyrinth replied:
수정 완료했습니다.

mskangg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,11 +1,35 @@
     private String contents;
-    private Long writerId;
+    @CreatedDate
+    @NotNull
```
javax.validation 의 @NotNull 은 jpa에서 동작하는 부분이 아닙니다.
`@Column(nullable = false, length = 6, updatable = false)` 을 추가해주세요!

OptimistLabyrinth replied:
수정 완료했습니다.

mskangg commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,25 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+
+@DataJpaTest
```
현재 `Answer` entity의 테스트를 전체적으로 추가해주세요.
`findByQuestionIdAndDeletedFalse, findByIdAndDeletedFalse 등등..`

나머지 entity들 모두 추가부탁드립니다.

OptimistLabyrinth replied:
수정 완료했습니다.

mskangg reviewed (changes requested):
안녕하세요 기성님!
jpa 미션 리뷰어를 맡은 강민수입니다.
잘부탁드려요 🙇‍♂️

1단계 미션 수행 잘 하셨습니다!
다만, 몇가지 충족되지 않아 피드백 드렸으니 반영 부탁드립니다.

전체적으로 Repository의 테스트코드 추가부탁드려요!🙇‍♂️

Commit:
feat: 기본생성자 public 너무 공개적으로 노출하지 않기

관련 링크
[https://github.com/next-step/jwp-qna/pull/617#discussion_r1014799076](https://github.com/next-step/jwp-qna/pull/617#discussion_r1014799076)

Commit:
feat: createdAt, updatedAt 컬럼에 spring data jpa auditing 사용하기

JPA 엔티티 컬럼의 nullable = false 속성에는
javax.validation.constraints.NotNull 를 사용하지 않음

관련 링크
[https://github.com/next-step/jwp-qna/pull/617#discussion_r1014799526](https://github.com/next-step/jwp-qna/pull/617#discussion_r1014799526)
[https://github.com/next-step/jwp-qna/pull/617#discussion_r1014799999](https://github.com/next-step/jwp-qna/pull/617#discussion_r1014799999)

Commit:
feat: User, Question, Answer 엔티티에 대한 테스트 코드 추가

XXXRepository 인터페이스에 있는 메서드 각각에 대한 테스트 케이스 추가
`@BeforeEach` 를 사용하기 위해서 `@Nested` 사용해서 테스트 코드 구조 개선
setDelete() 메서드 리턴 밸류 this 타입으로 변경

관련 링크
[https://github.com/next-step/jwp-qna/pull/617#discussion_r1014800717](https://github.com/next-step/jwp-qna/pull/617#discussion_r1014800717)

OptimistLabyrinth commented:
강민수님 반가워요 😄😆
2주차 JPA 미션 동안에 잘 부탁드립니다.

개별 도메인 엔티티에 대한 테스트 케이스를 추가했습니다.
Service 클래스에 대한 테스트 케이스도 추가하고 싶었지만, 
애플리케이션 세부적인 요구 조건에 대한 설명을 찾지 못해서 어떤 테스트 케이스가 필요한지 파악하지 못했어요.

서비스 클래스에 대한 테스트 코드도 필요한가요, 아니면 2단계에서 하게 되나요??

mskangg commented on src/main/java/qna/domain/BaseCreatedAndUpdatedAt.java:
```diff
@@ -0,0 +1,23 @@
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+@Getter
+public abstract class BaseCreatedAndUpdatedAt {
```
audit 적용하셨군요 👍

mskangg commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,11 +1,26 @@
+    @Lob
     private String contents;
-    private Long writerId;
+    @Column(nullable = false)
```
👍👍👍

mskangg commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,81 @@
+            final List<Answer> answers = new ArrayList<>(Arrays.asList(
+                    new Answer(user, questions.get(0), "answer content 1"),
+                    new Answer(user, questions.get(1), "answer content 2"),
+                    new Answer(user, questions.get(0), "answer content 3").setDeleted(true)
```
questions.get(2) 로 되어야할 것 같아요.
보통 이렇게 given 절에 테스트 객체를 만드는 경우 Fixtures 클래스를 선언하여 만들곤 합니다.

```java
public class Fixtures {

    public static List<Question> question2deletedQuestion1(User user){
        List<Question> questions = new ArrayList<>();
        questions.add(question("question title 1", "question content 1", user));
        questions.add(question("question title 2", "question content 2", user));
        questions.add(deletedQuestion("question title 3", "question content 3", user));
        return questions;
    }

    private static Question question(String title, String contents, User user) {
        return new Question(title, contents).writeBy(user);
    }

    private static Question deletedQuestion(String title, String contents, User user) {
        Question question = question(title, contents, user);
        question.setDeleted(true);
        return question;
    }
}
```

mskangg commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,81 @@
+            final List<Answer> foundAnswers =
+                    answerRepository.findByQuestionIdAndDeletedFalse(savedQuestions.get(0).getId());
+            assertAll(
+                    () -> assertThat(foundAnswers.size()).isEqualTo(1),
```
list는 사이즈 확인하기보단 hasSize 메서드를 사용해보아요 :)

OptimistLabyrinth replied:
오 hasSize 가 조금 더 직관적인 것 같아요.
조언 감사합니다 👍👍

mskangg reviewed (approved):
> 서비스 클래스에 대한 테스트 코드도 필요한가요, 아니면 2단계에서 하게 되나요??

서비스 클래스는 3단계 리팩토링 하기 전에 진행하시면 될 것 같아요!

전체적으로 피드백 반영 잘하셨습니다 👍 
소소한 피드백 남겨드렸으니 확인해보시고 적용필요하시면 다음 단계에 진행해주세요.
그럼 머지하겠습니다. 🚀
